thetis/callback.py

```diff
@@ -481,3 +481,43 @@ def __call__(self):
             field_vals.append(np.reshape(self._evaluate_field(field_name), (ndetectors, -1)))
 
         return np.hstack(field_vals)
+
+
+class FunctionalCallback(DiagnosticCallback):
+    """Callback that evaluates a functional involving integrals in both time and space. This callback can also be used
+    to assemble time dependent objective functionals for adjoint simulations. Time integration is achieved using the
+    trapezium rule."""
+    variable_names = ['current integral', 'objective value']
+
+    def __init__(self, scalar_callback, solver_obj, **kwargs):
+        """
+        :arg scalar_callback: Python function that returns a list of values of an objective functional.
+        :arg solver_obj: Thetis solver object
+        :arg **kwargs: any additional keyword arguments, see DiagnosticCallback
+        """
+        kwargs.setdefault('export_to_hdf5', False)
+        kwargs.setdefault('append_to_log', False)
+        super(FunctionalCallback, self).__init__(solver_obj, **kwargs)
+        self.scalar_callback = scalar_callback      # Evaluate functional
+        self.objective_value = []   # Store functional value
+        self.dt = solver_obj.options.timestep
+
+    def __call__(self):
+        value = self.scalar_callback()
```

Comment:
Here `value` is a numerical value (i.e. float), right?

Reply:
Yes, unless adjoint is being used such as described in the example above. If we use `from firedrake_adjoint import assemble`, rather than `from firedrake import assemble` then `value` is no longer a float.

Reply:
Even then, it's an AdjFloat - which acts like any other float (it's a subclass of). It just has additional annotation associated with it.

Reply:
Oh okay, I see! Thanks